File: backend-go/internal/modules/system/router.go
package system

import (
	_ "drip-admin/backend-go/docs"
	"drip-admin/backend-go/internal/config"
	"drip-admin/backend-go/internal/modules/system/controller"
	"drip-admin/backend-go/internal/modules/system/service"
	"github.com/gin-contrib/cors"
	"github.com/gin-gonic/gin"
	"github.com/redis/go-redis/v9"
	swaggerFiles "github.com/swaggo/files"
	ginSwagger "github.com/swaggo/gin-swagger"
	"go.uber.org/zap"
	"gorm.io/gorm"
)

type Server struct {
	cfg    config.Config
	db     *gorm.DB
	redis  *redis.Client
	logger *zap.Logger
}

func NewServer(cfg config.Config, db *gorm.DB, redisClient *redis.Client, logger *zap.Logger) *Server {
	return &Server{cfg: cfg, db: db, redis: redisClient, logger: logger}
}

func (s *Server) Router() *gin.Engine {
	gin.SetMode(gin.ReleaseMode)
	router := gin.New()
	svc := service.NewServer(s.cfg, s.db, s.redis, s.logger)
	router.Use(svc.RequestLoggerMiddleware(), svc.RecoveryMiddleware())
	router.Use(cors.New(cors.Config{
		AllowOriginFunc:  func(string) bool { return true },
		AllowMethods:     []string{"GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"},
		AllowHeaders:     []string{"*"},
		AllowCredentials: true,
	}))

	ctl := controller.New(svc)
	api := router.Group("/api")
	api.GET("/", ctl.Root)
	api.GET("/favicon.ico", ctl.Favicon)
	api.GET("/health", ctl.Health)
	api.GET("/v3/api-docs", ctl.OpenAPIDocs)
	api.GET("/swagger-ui.html", ctl.SwaggerUI)
	api.GET("/swagger-ui/*any", ginSwagger.WrapHandler(swaggerFiles.Handler, ginSwagger.URL("/api/v3/api-docs")))
	api.GET("/system/publicConfig", ctl.PublicConfig)
	api.POST("/system/login", ctl.Login)

	protected := api.Group("")
	protected.Use(svc.AuthMiddleware(), svc.OperationLogMiddleware())
	registerProtectedRoutes(protected, svc, ctl)
	return router
}

func registerProtectedRoutes(r *gin.RouterGroup, svc *service.Server, ctl *controller.Controller) {
	system := r.Group("/system")
	system.POST("/logout", ctl.Logout)
	system.GET("/me", ctl.Me)
	system.PUT("/password", ctl.Password)
	system.PUT("/profile", ctl.Profile)

	system.GET("/user", svc.RequirePermission("system:user:list"), ctl.Users)
	system.GET("/user/:id", svc.RequirePermission("system:user:detail"), ctl.User)
	system.POST("/user", svc.RequirePermission("system:user:create"), ctl.CreateUser)
	system.PUT("/user/:id", svc.RequirePermission("system:user:update"), ctl.UpdateUser)
	system.DELETE("/user/:id", svc.RequirePermission("system:user:delete"), ctl.DeleteUser)
	system.PUT("/user/:id/status", svc.RequirePermission("system:user:disable"), ctl.UserStatus)
	system.POST("/user/:id/unlock", svc.RequirePermission("system:user:unlock"), ctl.UnlockUser)
	system.PUT("/user/:id/role", svc.RequirePermission("system:user:assignRole"), ctl.UserRoles)
	system.POST("/user/:id/resetPassword", svc.RequirePermission("system:user:resetPassword"), ctl.ResetPassword)

	system.GET("/role", svc.RequirePermission("system:role:list"), ctl.Roles)
	system.GET("/role/option", svc.RequirePermission("system:role:list"), ctl.RoleOptions)
	system.GET("/role/:id", svc.RequirePermission("system:role:list"), ctl.Role)
	system.GET("/role/:id/user", svc.RequirePermission("system:role:list"), ctl.RoleUsers)
	system.GET("/role/:id/permission", svc.RequirePermission("system:role:permission"), ctl.RolePermissions)
	system.POST("/role", svc.RequirePermission("system:role:create"), ctl.CreateRole)
	system.PUT("/role/:id", svc.RequirePermission("system:role:update"), ctl.UpdateRole)
	system.DELETE("/role/:id", svc.RequirePermission("system:role:delete"), ctl.DeleteRole)
	system.PUT("/role/:id/status", svc.RequirePermission("system:role:update"), ctl.RoleStatus)
	system.PUT("/role/:id/permission", svc.RequirePermission("system:role:permission"), ctl.AssignRoleMenus)

	system.GET("/menu", svc.RequirePermission("system:menu:list"), ctl.Menus)
	system.POST("/menu", svc.RequirePermission("system:menu:create"), ctl.CreateMenu)
	system.PUT("/menu/:id", svc.RequirePermission("system:menu:update"), ctl.UpdateMenu)
	system.DELETE("/menu/:id", svc.RequirePermission("system:menu:delete"), ctl.DeleteMenu)
	system.PUT("/menu/:id/status", svc.RequirePermission("system:menu:status"), ctl.MenuStatus)

	system.GET("/dept", svc.RequirePermission("system:dept:list"), ctl.Depts)
	system.GET("/dept/:id", svc.RequirePermission("system:dept:list"), ctl.Dept)
	system.POST("/dept", svc.RequirePermission("system:dept:create"), ctl.CreateDept)
	system.PUT("/dept/:id", svc.RequirePermission("system:dept:update"), ctl.UpdateDept)
	system.DELETE("/dept/:id", svc.RequirePermission("system:dept:delete"), ctl.DeleteDept)
	system.PUT("/dept/:id/status", svc.RequirePermission("system:dept:update"), ctl.DeptStatus)

	system.GET("/config", svc.RequirePermission("system:config:list"), ctl.Configs)
	system.POST("/config", svc.RequirePermission("system:config:create"), ctl.CreateConfig)
	system.PUT("/config/:id", svc.RequirePermission("system:config:update"), ctl.UpdateConfig)
	system.DELETE("/config/:id", svc.RequirePermission("system:config:delete"), ctl.DeleteConfig)
	system.PUT("/config/:id/status", svc.RequirePermission("system:config:update"), ctl.ConfigStatus)

	system.GET("/dict/type", svc.RequirePermission("system:dict:list"), ctl.DictTypes)
	system.POST("/dict/type", svc.RequirePermission("system:dict:create"), ctl.CreateDictType)
	system.PUT("/dict/type/:id", svc.RequirePermission("system:dict:update"), ctl.UpdateDictType)
	system.DELETE("/dict/type/:id", svc.RequirePermission("system:dict:delete"), ctl.DeleteDictType)
	system.GET("/dict/type/:id/item", svc.RequirePermission("system:dict:list"), ctl.DictItems)
	system.POST("/dict/item", svc.RequirePermission("system:dict:create"), ctl.CreateDictItem)
	system.PUT("/dict/item/:id", svc.RequirePermission("system:dict:update"), ctl.UpdateDictItem)
	system.DELETE("/dict/item/:id", svc.RequirePermission("system:dict:delete"), ctl.DeleteDictItem)
	system.PUT("/dict/item/:id/status", svc.RequirePermission("system:dict:update"), ctl.DictItemStatus)
	system.POST("/dict/cache/refresh", svc.RequirePermission("system:dict:update"), ctl.RefreshDictCache)

	system.GET("/onlineUser", svc.RequirePermission("system:online:list"), ctl.OnlineUsers)
	system.GET("/onlineUser/:tokenId", svc.RequirePermission("system:online:list"), ctl.OnlineUser)
	system.POST("/onlineUser/:tokenId/kickout", svc.RequirePermission("system:online:kickout"), ctl.KickoutOnlineUser)

	system.GET("/loginLog", svc.RequirePermission("system:loginLog:list"), ctl.LoginLogs)
	system.GET("/loginLog/:id", svc.RequirePermission("system:loginLog:list"), ctl.LoginLog)
	system.GET("/operationLog", svc.RequirePermission("system:operationLog:list"), ctl.OperationLogs)
	system.GET("/operationLog/:id", svc.RequirePermission("system:operationLog:list"), ctl.OperationLog)

	system.GET("/job", svc.RequirePermission("system:job:list"), ctl.Jobs)
	system.GET("/job/scripts", svc.RequirePermission("system:job:list"), ctl.JobScripts)
	system.GET("/job/:id", svc.RequirePermission("system:job:list"), ctl.Job)
	system.POST("/job", svc.RequirePermission("system:job:create"), ctl.CreateJob)
	system.PUT("/job/:id", svc.RequirePermission("system:job:update"), ctl.UpdateJob)
	system.DELETE("/job/:id", svc.RequirePermission("system:job:delete"), ctl.DeleteJob)
	system.PUT("/job/:id/status", svc.RequirePermission("system:job:update"), ctl.JobStatus)
	system.POST("/job/:id/run", svc.RequirePermission("system:job:run"), ctl.RunJob)
	system.GET("/job/:id/runLog", svc.RequirePermission("system:job:list"), ctl.JobLogs)
	system.GET("/jobRunLog", svc.RequirePermission("system:job:history"), ctl.JobRunLogs)

	system.POST("/files", svc.RequirePermission("system:file:upload"), ctl.UploadFile)

	printTemplate := system.Group("/print-template")
	printTemplate.GET("", svc.RequirePermission("system:printTemplate:list"), ctl.PrintTemplates)
	printTemplate.GET("/:id", svc.RequirePermission("system:printTemplate:list"), ctl.PrintTemplate)
	printTemplate.POST("", svc.RequirePermission("system:printTemplate:create"), ctl.CreatePrintTemplate)
	printTemplate.POST("/:id/copy", svc.RequirePermission("system:printTemplate:create"), ctl.CopyPrintTemplate)
	printTemplate.PUT("/:id", svc.RequirePermission("system:printTemplate:update"), ctl.UpdatePrintTemplate)
	printTemplate.DELETE("/:id", svc.RequirePermission("system:printTemplate:delete"), ctl.DeletePrintTemplate)
	printTemplate.PUT("/:id/status", svc.RequirePermission("system:printTemplate:update"), ctl.PrintTemplateStatus)
}

File: backend-rust/src/modules/system/router.rs
use super::AppState;
use super::controller::{
    auth_controller, common_controller, config_controller, dept_controller, dict_controller,
    file_controller, job_controller, log_controller, menu_controller, online_user_controller,
    print_template_controller, role_controller, user_controller,
};
use super::service::permission_service::{auth_required, require_permission};
use super::service::request_log_service::record_request;
use axum::Router;
use axum::middleware;
use axum::routing::{delete, get, post, put};
use tower_http::cors::{Any, CorsLayer};
use tower_http::trace::TraceLayer;
use utoipa_swagger_ui::SwaggerUi;

pub fn create_router(state: AppState) -> Router {
    let cors = CorsLayer::new()
        .allow_origin(Any)
        .allow_methods(Any)
        .allow_headers(Any);

    let public = Router::new()
        .route("/", get(common_controller::root))
        .route("/favicon.ico", get(common_controller::favicon))
        .route("/health", get(common_controller::health))
        .route(
            "/system/publicConfig",
            get(config_controller::public_config),
        )
        .route("/system/login", post(auth_controller::login))
        .merge(SwaggerUi::new("/swagger-ui").url("/v3/api-docs", super::api_doc::openapi()));

    let protected = Router::new()
        .route("/system/logout", post(auth_controller::logout))
        .route("/system/me", get(auth_controller::me))
        .route("/system/password", put(auth_controller::password))
        .route("/system/profile", put(auth_controller::profile))
        .route(
            "/system/user",
            get(user_controller::list).route_layer(require_permission("system:user:list")),
        )
        .route(
            "/system/user/{id}",
            get(user_controller::detail).route_layer(require_permission("system:user:detail")),
        )
        .route(
            "/system/user",
            post(user_controller::create).route_layer(require_permission("system:user:create")),
        )
        .route(
            "/system/user/{id}",
            put(user_controller::update).route_layer(require_permission("system:user:update")),
        )
        .route(
            "/system/user/{id}",
            delete(user_controller::delete).route_layer(require_permission("system:user:delete")),
        )
        .route(
            "/system/user/{id}/status",
            put(user_controller::status).route_layer(require_permission("system:user:disable")),
        )
        .route(
            "/system/user/{id}/unlock",
            post(user_controller::unlock).route_layer(require_permission("system:user:unlock")),
        )
        .route(
            "/system/user/{id}/role",
            put(user_controller::assign_roles)
                .route_layer(require_permission("system:user:assignRole")),
        )
        .route(
            "/system/user/{id}/resetPassword",
            post(user_controller::reset_password)
                .route_layer(require_permission("system:user:resetPassword")),
        )
        .merge(role_routes())
        .merge(menu_routes())
        .merge(dept_routes())
        .merge(config_routes())
        .merge(dict_routes())
        .merge(online_user_routes())
        .merge(log_routes())
        .merge(job_routes())
        .merge(file_routes())
        .merge(print_template_routes())
        .route_layer(middleware::from_fn_with_state(state.clone(), auth_required));

    Router::new()
        .route("/api/", get(common_controller::root))
        .nest("/api", public.merge(protected))
        .layer(middleware::from_fn_with_state(state.clone(), record_request))
        .layer(cors)
        .layer(TraceLayer::new_for_http())
        .with_state(state)
}

fn role_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/role",
            get(role_controller::list).route_layer(require_permission("system:role:list")),
        )
        .route(
            "/system/role/option",
            get(role_controller::options).route_layer(require_permission("system:role:list")),
        )
        .route(
            "/system/role/{id}",
            get(role_controller::detail).route_layer(require_permission("system:role:list")),
        )
        .route(
            "/system/role/{id}/user",
            get(role_controller::users).route_layer(require_permission("system:role:list")),
        )
        .route(
            "/system/role/{id}/permission",
            get(role_controller::permissions)
                .route_layer(require_permission("system:role:permission")),
        )
        .route(
            "/system/role",
            post(role_controller::create).route_layer(require_permission("system:role:create")),
        )
        .route(
            "/system/role/{id}",
            put(role_controller::update).route_layer(require_permission("system:role:update")),
        )
        .route(
            "/system/role/{id}",
            delete(role_controller::delete).route_layer(require_permission("system:role:delete")),
        )
        .route(
            "/system/role/{id}/status",
            put(role_controller::status).route_layer(require_permission("system:role:update")),
        )
        .route(
            "/system/role/{id}/permission",
            put(role_controller::assign_permissions)
                .route_layer(require_permission("system:role:permission")),
        )
}

fn menu_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/menu",
            get(menu_controller::list).route_layer(require_permission("system:menu:list")),
        )
        .route(
            "/system/menu",
            post(menu_controller::create).route_layer(require_permission("system:menu:create")),
        )
        .route(
            "/system/menu/{id}",
            put(menu_controller::update).route_layer(require_permission("system:menu:update")),
        )
        .route(
            "/system/menu/{id}",
            delete(menu_controller::delete).route_layer(require_permission("system:menu:delete")),
        )
        .route(
            "/system/menu/{id}/status",
            put(menu_controller::status).route_layer(require_permission("system:menu:status")),
        )
}

fn dept_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/dept",
            get(dept_controller::list).route_layer(require_permission("system:dept:list")),
        )
        .route(
            "/system/dept/{id}",
            get(dept_controller::detail).route_layer(require_permission("system:dept:list")),
        )
        .route(
            "/system/dept",
            post(dept_controller::create).route_layer(require_permission("system:dept:create")),
        )
        .route(
            "/system/dept/{id}",
            put(dept_controller::update).route_layer(require_permission("system:dept:update")),
        )
        .route(
            "/system/dept/{id}",
            delete(dept_controller::delete).route_layer(require_permission("system:dept:delete")),
        )
        .route(
            "/system/dept/{id}/status",
            put(dept_controller::status).route_layer(require_permission("system:dept:update")),
        )
}

fn config_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/config",
            get(config_controller::list).route_layer(require_permission("system:config:list")),
        )
        .route(
            "/system/config",
            post(config_controller::create).route_layer(require_permission("system:config:create")),
        )
        .route(
            "/system/config/{id}",
            get(config_controller::detail).route_layer(require_permission("system:config:list")),
        )
        .route(
            "/system/config/{id}",
            put(config_controller::update).route_layer(require_permission("system:config:update")),
        )
        .route(
            "/system/config/{id}",
            delete(config_controller::delete)
                .route_layer(require_permission("system:config:delete")),
        )
        .route(
            "/system/config/{id}/status",
            put(config_controller::status).route_layer(require_permission("system:config:update")),
        )
}

fn dict_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/dict/type",
            get(dict_controller::type_list).route_layer(require_permission("system:dict:list")),
        )
        .route(
            "/system/dict/type",
            post(dict_controller::create_type)
                .route_layer(require_permission("system:dict:create")),
        )
        .route(
            "/system/dict/type/{id}",
            put(dict_controller::update_type).route_layer(require_permission("system:dict:update")),
        )
        .route(
            "/system/dict/type/{id}",
            delete(dict_controller::delete_type)
                .route_layer(require_permission("system:dict:delete")),
        )
        .route(
            "/system/dict/type/{id}/item",
            get(dict_controller::items).route_layer(require_permission("system:dict:list")),
        )
        .route(
            "/system/dict/item",
            post(dict_controller::create_item)
                .route_layer(require_permission("system:dict:create")),
        )
        .route(
            "/system/dict/item/{id}",
            put(dict_controller::update_item).route_layer(require_permission("system:dict:update")),
        )
        .route(
            "/system/dict/item/{id}",
            delete(dict_controller::delete_item)
                .route_layer(require_permission("system:dict:delete")),
        )
        .route(
            "/system/dict/item/{id}/status",
            put(dict_controller::item_status).route_layer(require_permission("system:dict:update")),
        )
        .route(
            "/system/dict/cache/refresh",
            post(dict_controller::refresh_cache)
                .route_layer(require_permission("system:dict:update")),
        )
}

fn online_user_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/onlineUser",
            get(online_user_controller::list).route_layer(require_permission("system:online:list")),
        )
        .route(
            "/system/onlineUser/{tokenId}",
            get(online_user_controller::detail)
                .route_layer(require_permission("system:online:list")),
        )
        .route(
            "/system/onlineUser/{tokenId}/kickout",
            post(online_user_controller::kickout)
                .route_layer(require_permission("system:online:kickout")),
        )
}

fn log_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/loginLog",
            get(log_controller::login_logs).route_layer(require_permission("system:loginLog:list")),
        )
        .route(
            "/system/loginLog/{id}",
            get(log_controller::login_log).route_layer(require_permission("system:loginLog:list")),
        )
        .route(
            "/system/operationLog",
            get(log_controller::operation_logs)
                .route_layer(require_permission("system:operationLog:list")),
        )
        .route(
            "/system/operationLog/{id}",
            get(log_controller::operation_log)
                .route_layer(require_permission("system:operationLog:list")),
        )
}

fn job_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/job",
            get(job_controller::list).route_layer(require_permission("system:job:list")),
        )
        .route(
            "/system/job/scripts",
            get(job_controller::scripts).route_layer(require_permission("system:job:list")),
        )
        .route(
            "/system/job/{id}",
            get(job_controller::detail).route_layer(require_permission("system:job:list")),
        )
        .route(
            "/system/job",
            post(job_controller::create).route_layer(require_permission("system:job:create")),
        )
        .route(
            "/system/job/{id}",
            put(job_controller::update).route_layer(require_permission("system:job:update")),
        )
        .route(
            "/system/job/{id}",
            delete(job_controller::delete).route_layer(require_permission("system:job:delete")),
        )
        .route(
            "/system/job/{id}/status",
            put(job_controller::status).route_layer(require_permission("system:job:update")),
        )
        .route(
            "/system/job/{id}/run",
            post(job_controller::run).route_layer(require_permission("system:job:run")),
        )
        .route(
            "/system/job/{id}/runLog",
            get(job_controller::run_logs).route_layer(require_permission("system:job:list")),
        )
        .route(
            "/system/jobRunLog",
            get(job_controller::all_run_logs).route_layer(require_permission("system:job:history")),
        )
}

fn file_routes() -> Router<AppState> {
    Router::new().route(
        "/system/files",
        post(file_controller::upload).route_layer(require_permission("system:file:upload")),
    )
}

fn print_template_routes() -> Router<AppState> {
    Router::new()
        .route(
            "/system/print-template",
            get(print_template_controller::list)
                .route_layer(require_permission("system:printTemplate:list")),
        )
        .route(
            "/system/print-template/{id}",
            get(print_template_controller::detail)
                .route_layer(require_permission("system:printTemplate:list")),
        )
        .route(
            "/system/print-template",
            post(print_template_controller::create)
                .route_layer(require_permission("system:printTemplate:create")),
        )
        .route(
            "/system/print-template/{id}/copy",
            post(print_template_controller::copy)
                .route_layer(require_permission("system:printTemplate:create")),
        )
        .route(
            "/system/print-template/{id}",
            put(print_template_controller::update)
                .route_layer(require_permission("system:printTemplate:update")),
        )
        .route(
            "/system/print-template/{id}",
            delete(print_template_controller::delete)
                .route_layer(require_permission("system:printTemplate:delete")),
        )
        .route(
            "/system/print-template/{id}/status",
            put(print_template_controller::status)
                .route_layer(require_permission("system:printTemplate:update")),
        )
}
